File: search/metrics.go
package search

import (
	"fmt"
	"sort"
	"strconv"
	"strings"
	"sync"
	"time"
)

const (
	httpStatusOK         = 200
	httpStatusBadRequest = 400
)

var latencyBuckets = []float64{0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1, 2.5, 5, 10}

type Metrics struct {
	mu         sync.Mutex
	requests   uint64
	results    uint64
	status     map[int]uint64
	latency    []uint64
	latencySum float64
}

func NewMetrics() *Metrics {
	return &Metrics{status: make(map[int]uint64), latency: make([]uint64, len(latencyBuckets)+1)}
}

func (m *Metrics) Observe(status, results int, duration time.Duration) {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.requests++
	m.results += uint64(results)
	m.status[status]++
	seconds := duration.Seconds()
	m.latencySum += seconds
	index := sort.SearchFloat64s(latencyBuckets, seconds)
	m.latency[index]++
}

func (m *Metrics) Prometheus() string {
	m.mu.Lock()
	defer m.mu.Unlock()
	var output strings.Builder
	fmt.Fprintf(&output, "# TYPE photon_search_requests_total counter\nphoton_search_requests_total %d\n", m.requests)
	statuses := make([]int, 0, len(m.status))
	for status := range m.status {
		statuses = append(statuses, status)
	}
	sort.Ints(statuses)
	for _, status := range statuses {
		fmt.Fprintf(&output, "photon_search_requests_status_total{status=\"%s\"} %d\n", strconv.Itoa(status), m.status[status])
	}
	fmt.Fprintf(&output, "# TYPE photon_search_results_total counter\nphoton_search_results_total %d\n", m.results)
	fmt.Fprintf(&output, "# TYPE photon_search_request_duration_seconds histogram\n")
	var cumulative uint64
	for index, bucket := range latencyBuckets {
		cumulative += m.latency[index]
		fmt.Fprintf(&output, "photon_search_request_duration_seconds_bucket{le=\"%g\"} %d\n", bucket, cumulative)
	}
	cumulative += m.latency[len(latencyBuckets)]
	fmt.Fprintf(&output, "photon_search_request_duration_seconds_bucket{le=\"+Inf\"} %d\nphoton_search_request_duration_seconds_sum %g\nphoton_search_request_duration_seconds_count %d\n", cumulative, m.latencySum, m.requests)
	return output.String()
}

File: search/src/metrics.py
class Metrics:
    def __init__(self):
        self.requests = 0
        self.results = 0
        self.status = {}
        self.latencies = []
        self.retrieval_modes = {}

    def observe(self, status, results, duration_seconds=0, retrieval="dense"):
        self.requests += 1
        self.results += results
        self.status[status] = self.status.get(status, 0) + 1
        self.latencies.append(duration_seconds)
        self.retrieval_modes[retrieval] = self.retrieval_modes.get(retrieval, 0) + 1

    def render(self):
        buckets = (0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1, 2.5, 5, 10)
        lines = [f"photon_search_requests_total {self.requests}", f"photon_search_results_total {self.results}"]
        lines.extend(f'photon_search_requests_status_total{{status="{status}"}} {count}' for status, count in sorted(self.status.items()))
        lines.extend(f'photon_search_retrieval_requests_total{{mode="{mode}"}} {count}' for mode, count in sorted(self.retrieval_modes.items()))
        lines.append("# TYPE photon_search_request_duration_seconds histogram")
        for bucket in buckets:
            lines.append(f'photon_search_request_duration_seconds_bucket{{le="{bucket:g}"}} {sum(value <= bucket for value in self.latencies)}')
        lines.extend([
            f"photon_search_request_duration_seconds_bucket{{le=\"+Inf\"}} {len(self.latencies)}",
            f"photon_search_request_duration_seconds_sum {sum(self.latencies)}",
            f"photon_search_request_duration_seconds_count {len(self.latencies)}",
        ])
        return "\n".join(lines) + "\n"
